app/src/main/java/com/android/streetworkapp/ui/profile/Profile.kt

```diff
@@ -33,47 +41,130 @@ fun ProfileScreen(
-          Button(
-              onClick = { navigationActions.navigateTo(Screen.ADD_FRIEND) },
-              modifier = Modifier.size(220.dp, 50.dp).testTag("profileAddButton")) {
-                Text(text = "Add a new friend", fontSize = 17.sp)
-              }
+                  // button to add a new friend
+                  Button(
+                      onClick = { navigationActions.navigateTo(Screen.ADD_FRIEND) },
+                      modifier = Modifier.size(220.dp, 50.dp).testTag("profileAddButton")) {
+                        Text(text = "Add a new friend", fontSize = 17.sp)
+                      }
 
-          Spacer(modifier = Modifier.height(10.dp))
-          // button to train with a friend
-          Button(
-              onClick = {},
-              colors = ButtonDefaults.buttonColors(Color(0xFFA53A36)),
-              modifier = Modifier.size(220.dp, 50.dp).testTag("profileTrainButton")) {
-                Text(text = "Train with a friend", fontSize = 17.sp)
+                  Spacer(modifier = Modifier.height(10.dp).testTag("profileInfoSpacer3"))
+
+                  // button to train with a friend
+                  Button(
+                      onClick = {
+                        Toast.makeText(context, "Not implemented yet", Toast.LENGTH_LONG).show()
+                      },
+                      colors = ButtonDefaults.buttonColors(Color(0xFFA53A36)),
+                      modifier = Modifier.size(220.dp, 50.dp).testTag("profileTrainButton")) {
+                        Text(text = "Train with a friend", fontSize = 17.sp)
+                      }
+                }
               }
+
+          DisplayFriendList(friendList, innerPaddingValues)
         }
   }
 }
+
+@Composable
+fun DisplayFriendList(friends: List<User>, padding: PaddingValues = PaddingValues(0.dp)) {
+  return if (friends.isNotEmpty()) {
+
+    // LazyColumn is scrollable
+    LazyColumn(
+        contentPadding = PaddingValues(vertical = 2.dp),
+        modifier =
+            Modifier.fillMaxWidth()
+                .padding(horizontal = 16.dp)
+                .padding(padding)
+                .testTag("friendList")) {
+          items(friends) { friend -> DisplayFriend(friend, padding) }
+        }
+  } else {
+    Text(
+        modifier = Modifier.padding(padding).testTag("emptyFriendListText"),
+        fontSize = 20.sp,
+        text = "You have no friends yet :(")
+  }
+}
+
+@Composable
+fun DisplayFriend(friend: User, padding: PaddingValues = PaddingValues(0.dp)) {
+  return Row(modifier = Modifier.fillMaxWidth().padding(padding).testTag("friendRow")) {
+
+    // profile placeholder
+    Image(
+        painter = painterResource(id = R.drawable.profile),
+        contentDescription = "profile picture",
+        modifier = Modifier.size(75.dp).testTag("friendProfilePicture"))
+
+    Spacer(modifier = Modifier.width(10.dp))
+
+    Column(modifier = Modifier.fillMaxWidth().padding(padding)) {
+
+      // Small spacing between elements
```

Comment:
Refactor: I do not obtain the same result as you in the UI. The spacing (at least in my laptop) between the friends are way too large. Here is a version that gives a more convinient spacing between friends:

(in summary, you should remove all the paddings and remove some of the spacers)

```
@Composable
fun DisplayFriendList(friends: List<User>, padding: PaddingValues = PaddingValues(0.dp)) {
  return if (friends.isNotEmpty()) {
    // LazyColumn is scrollable
    LazyColumn(
        contentPadding = PaddingValues(vertical = 0.dp),
        modifier =
            Modifier.fillMaxWidth()
                .testTag("friendList")) {
          items(friends) { friend -> DisplayFriend(friend) }
        }
  } else {
    Text(
        modifier = Modifier.padding(padding).testTag("emptyFriendListText"),
        fontSize = 20.sp,
        text = "You have no friends yet :(")
  }
}

```

```
@Composable
fun DisplayFriend(friend: User) {
  Row(modifier = Modifier.fillMaxWidth().testTag("friendRow")) {
    // profile placeholder
    Image(
        painter = painterResource(id = R.drawable.profile),
        contentDescription = "profile picture",
        modifier = Modifier.size(75.dp).testTag("friendProfilePicture"))

    Column(modifier = Modifier.fillMaxWidth()) {

      Text(fontSize = 22.sp, text = friend.username, modifier = Modifier.testTag("friendUsername"))

      // score
      Text(
          text = "Score: ${friend.score}",
          fontSize = 22.sp,
          modifier = Modifier.testTag("friendScore"))
    }
  }
}


```


Please note that those are suggestion based on the current code. This could be improved in other way if you wish, but it is important to change the way friends are displayed

Reply:
I am unsure what kind of device you use. I removed all padding and kept only one spacer which prevents the friend username from being too high relatively to the image. Is it better like this for you ?